// cognitive-services/dist/search/bingImageSearch.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var commonService = require('../commonService');

/**
 * The Image Search API lets you send a search query to Bing and get back a list of relevant images.
 * @augments commonService
 * {@link https://docs.microsoft.com/en-us/rest/api/cognitiveservices/bing-images-api-v5-reference|documentation}
 */

var bingImageSearch = function (_commonService) {
    _inherits(bingImageSearch, _commonService);

    /**
     * Constructor.
     * 
     * @param {Object} obj
     * @param {string} obj.apiKey
     */
    function bingImageSearch(_ref) {
        var apiKey = _ref.apiKey;

        _classCallCheck(this, bingImageSearch);

        var endpoint = "api.cognitive.microsoft.com";

        var _this = _possibleConstructorReturn(this, _commonService.call(this, { apiKey: apiKey, endpoint: endpoint }));

        _this.endpoints = [endpoint];
        return _this;
    }

    /**
     * Get insights for an image sent in the POST body.
     * @returns {Promise.<object>}
     */


    bingImageSearch.prototype.imageInsights = function imageInsights(_ref2) {
        var parameters = _ref2.parameters,
            body = _ref2.body;


        var operation = {
            "path": "bing/v5.0/images/search",
            "method": "POST",
            "operationId": "571fab09dbe2d933e891028f",
            "parameters": [{
                "name": "q",
                "description": "The user's search query string",
                "value": null,
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters,
            body: body
        });
    };

    /**
     * Get relevant images for a given query.
     * @returns {Promise.<object>}
     */
    bingImageSearch.prototype.search = function search(_ref3) {
        var parameters = _ref3.parameters;


        var operation = {
            "path": "bing/v5.0/images/search",
            "method": "GET",
            "operationId": "56b4433fcf5ff8098cef380c",
            "parameters": [{
                "name": "q",
                "description": "The user's search query string",
                "value": "",
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "count",
                "description": "The number of image results to return in the response. The actual number delivered may be less than requested.",
                "value": "",
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "offset",
                "description": "The zero-based offset that indicates the number of image results to skip before returning results.",
                "value": "",
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "mkt",
                "description": "The market where the results come from. Typically, this is the country where the user is making the request from; however, it could be a different country if the user is not located in a country where Bing delivers results. The market must be in the form {language code}-{country code}. For example, en-US.\n\n<br>\n<br>\nFull list of supported markets:\n<br>\nes-AR,en-AU,de-AT,nl-BE,fr-BE,pt-BR,en-CA,fr-CA,es-CL,da-DK,fi-FI,fr-FR,de-DE,zh-HK,en-IN,en-ID,en-IE,it-IT,ja-JP,ko-KR,en-MY,es-MX,nl-NL,en-NZ,no-NO,zh-CN,pl-PL,pt-PT,en-PH,ru-RU,ar-SA,en-ZA,es-ES,sv-SE,fr-CH,de-CH,zh-TW,tr-TR,en-GB,en-US,es-US",
                "value": "en-us",
                "required": false,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "safeSearch",
                "description": "A filter used to filter results for adult content.",
                "value": "Off",
                "options": ["Off", "Moderate", "Strict"],
                "required": false,
                "type": "queryStringParam",
                "typeName": "string"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters
        });
    };

    /**
     * Get currently trending images.
     * @returns {Promise.<object>}
     */
    bingImageSearch.prototype.trending = function trending() {

        var operation = {
            "path": "bing/v5.0/images/trending",
            "method": "GET",
            "operationId": "56b44b8ccf5ff81038d15ce0"
        };

        return this.makeRequest({
            operation: operation
        });
    };

    return bingImageSearch;
}(commonService);

module.exports = bingImageSearch;
