// cognitive-services/dist/language/webLanguageModel.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var commonService = require('../commonService');

/**
 * Using this API, your application can leverage the power of big data through language models trained on web-scale corpora collected by Bing in the EN-US market.
 * 
 * These smoothed backoff N-gram language models, supporting Markov order up to 5, are trained on the following corpora:
    - Web page body text
    - Web page title text
    - Web page anchor text
    - Web search query text

The Web LM REST API supports four lookup operations:

- Joint (log10) probability of a sequence of words.
- Conditional (log10) probability of one word given a sequence of preceding words.
- List of words (completions) most likely to follow a given sequence of words.
- Word breaking of strings that contain no spaces.
 * @augments commonService
 * {@link https://westus.dev.cognitive.microsoft.com/docs/services/55de9ca4e597ed1fd4e2f104/operations/55de9ca4e597ed19b0de8a51|documentation}
 */

var webLanguageModel = function (_commonService) {
    _inherits(webLanguageModel, _commonService);

    /**
     * Constructor.
     * 
     * @param {Object} obj
     * @param {string} obj.apiKey
     * @param {string} obj.endpoint
     */
    function webLanguageModel(_ref) {
        var apiKey = _ref.apiKey,
            endpoint = _ref.endpoint;

        _classCallCheck(this, webLanguageModel);

        var _this = _possibleConstructorReturn(this, _commonService.call(this, { apiKey: apiKey, endpoint: endpoint }));

        _this.endpoints = ["westus.api.cognitive.microsoft.com"];
        return _this;
    }

    /**
     * Insert spaces into a string of words lacking spaces, like a hashtag or part of a URL. Punctuation or exotic characters can prevent a string from being broken, so it’s best to limit input strings to lower-case, alpha-numeric characters.
    Example Parameters: {
        "model": null,
        "text": null,
        "order": null,
        "maxNumOfCandidatesReturned": null
    }
    @returns {Promise.<object>}
    */


    webLanguageModel.prototype.breakIntoWords = function breakIntoWords(_ref2) {
        var parameters = _ref2.parameters;


        var operation = {
            "path": "text/weblm/v1.0/breakIntoWords",
            "method": "POST",
            "operationId": "55de9ca4e597ed19b0de8a51",
            "parameters": [{
                "name": "model",
                "description": "Which model to use.",
                "value": null,
                "options": ["title", "anchor", "query", "body"],
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "text",
                "description": "The line of text to break into words. If spaces are present, they will be interpreted as hard breaks and maintained, except for leading or trailing spaces, which will be trimmed.",
                "value": null,
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "order",
                "description": "The order of N-gram.",
                "value": "5",
                "options": ["1", "2", "3", "4", "5"],
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "maxNumOfCandidatesReturned",
                "description": "Max number of candidates that would be returned. If not specified, use default value 5.",
                "value": "5",
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters
        });
    };

    /**
     * Calculate the conditional probability that a particular word will follow a given sequence of words.
    Example Parameters: {
        "model": null,
        "order": null
    }
    @returns {Promise.<object>}
    */
    webLanguageModel.prototype.calculateConditionalProbability = function calculateConditionalProbability(_ref3) {
        var parameters = _ref3.parameters,
            body = _ref3.body;


        var operation = {
            "path": "text/weblm/v1.0/calculateConditionalProbability",
            "method": "POST",
            "operationId": "55de9ca4e597ed19b0de8a4e",
            "parameters": [{
                "name": "model",
                "description": "Which model to use.",
                "value": null,
                "options": ["title", "anchor", "query", "body"],
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "order",
                "description": "The order of N-gram.",
                "value": "5",
                "options": ["1", "2", "3", "4", "5"],
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "queries",
                "description": "Array of queries",
                "value": null,
                "required": false,
                "type": "inBody",
                "typeName": "array"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters,
            headers: { "Content-type": "application/json" },
            body: body
        });
    };

    /**
     * Calculate the joint probability that a particular sequence of words will appear together.
    Example Parameters: {
        "model": null,
        "order": null
    }
    @returns {Promise.<object>}
    */
    webLanguageModel.prototype.calculateJointProbability = function calculateJointProbability(_ref4) {
        var parameters = _ref4.parameters,
            body = _ref4.body;


        var operation = {
            "path": "text/weblm/v1.0/calculateJointProbability",
            "method": "POST",
            "operationId": "55de9ca4e597ed19b0de8a4f",
            "parameters": [{
                "name": "model",
                "description": "Which model to use.",
                "value": null,
                "options": ["title", "anchor", "query", "body"],
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "order",
                "description": "The order of N-gram.",
                "value": "5",
                "options": ["1", "2", "3", "4", "5"],
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "queries",
                "description": "Array of queries",
                "value": null,
                "required": false,
                "type": "inBody",
                "typeName": "array"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters,
            headers: { "Content-type": "application/json" },
            body: body
        });
    };

    /**
     * Get the list of words (completions) most likely to follow a given sequence of words.
    Example Parameters: {
        "model": null,
        "words": null,
        "order": null,
        "maxNumOfCandidatesReturned": null
    }
    @returns {Promise.<object>}
    */
    webLanguageModel.prototype.generateNextWords = function generateNextWords(_ref5) {
        var parameters = _ref5.parameters;


        var operation = {
            "path": "text/weblm/v1.0/generateNextWords",
            "method": "POST",
            "operationId": "55de9ca4e597ed19b0de8a50",
            "parameters": [{
                "name": "model",
                "description": "Which model to use.",
                "value": null,
                "options": ["title", "anchor", "query", "body"],
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "words",
                "description": "A string containing a sequence of words from which to generate the list of words likely to follow. The words should be separated by spaces.",
                "value": null,
                "required": true,
                "type": "queryStringParam",
                "typeName": "string"
            }, {
                "name": "order",
                "description": "The order of N-gram.",
                "value": "5",
                "options": ["1", "2", "3", "4", "5"],
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }, {
                "name": "maxNumOfCandidatesReturned",
                "description": "Max number of candidates that would be returned. If not specified, use default value 5.",
                "value": "5",
                "required": false,
                "type": "queryStringParam",
                "typeName": "number"
            }]
        };

        return this.makeRequest({
            operation: operation,
            parameters: parameters
        });
    };

    /**
     * List models available currently.
     * @returns {Promise.<object>}
    */
    webLanguageModel.prototype.listAvailableModels = function listAvailableModels() {

        var operation = {
            "path": "text/weblm/v1.0/models",
            "method": "GET",
            "operationId": "565bf87b778daf12447f43c1"
        };

        return this.makeRequest({
            operation: operation
        });
    };

    return webLanguageModel;
}(commonService);

;

module.exports = webLanguageModel;
